js-packages/@prestojs/routing/src/useUrlQueryState.ts

```diff
@@ -0,0 +1,447 @@
+ *
+ * ```jslive
+ * # OPTIONS: {"fakeBrowser": true}
+ * function ExampleUrlSync() {
+ *   const parse = (c) => Number(c);
+ *   const [urlState, setUrlState] = useUrlQueryState(
+ *     { count: 1 },
+ *     { parse, prefix: 'test_' }
+ *   );
+ *   const onClick = () => setUrlState(s => ({ count: s.count + 1 }));
+ *   return (
+ *     <Button onClick={onClick}>{urlState.count} Increment</Button>
+ *   );
+ * }
+ * <ExampleUrlSync prefix="p_" />
+ * ```
+ *
+ * Partially control URL query params when no prefix is set
+ * ```jslive
+ * # OPTIONS: {"fakeBrowser": "/?name=test&page=1&count=5"}
+ * function ExampleUrlSyncControlled() {
+ *   const parse = (c) => Number(c);
+ *   const [urlState, setUrlState] = useUrlQueryState(
+ *     { count: 1 },
+ *     { parse, controlledKeys: ['count'] }
+ *   );
+ *   const onClick = () => setUrlState(s => ({ count: s.count + 1 }));
+ *   return (
+ *     <Button onClick={onClick}>{urlState.count} Increment</Button>
+ *   );
+ * }
+ * <ExampleUrlSyncControlled />
+ * ```
+ *
+ * Advanced parse & stringify
+ *
+ * ```jslive
+ * # OPTIONS: {"fakeBrowser": true }
+ * function ExampleUrlSyncControlled() {
+ *   const parse = (value, key) => {
+ *       if (key === 'count') {
+ *           return Number(value);
+ *       }
+ *       if (key === 'data') {
+ *           return JSON.parse(value);
+ *       }
+ *       return value;
+ *   }
+ *   const stringify = (value, key) => {
+ *       if (key === 'data') {
+ *           return JSON.stringify(value);
+ *       }
+ *       return value;
+ *   }
+ *   const [urlState, setUrlState] = useUrlQueryState(
+ *     { data: { name: 'Dave', email: '' }, count: 1 },
+ *     { parse, stringify }
+ *   );
+ *   const onClick = () => setUrlState(s => ({ ...s, count: s.count + 1 }));
+ *   const { data = {} } = urlState;
+ *   return (
+ *     <div>
+ *        <input
+ *          value={data.name}
+ *          onChange={({ target: { value }}) => setUrlState(s => ({
+ *            ...s,
+ *            data: { ...s.data, name: value },
+ *          }))}
+ *        />
+ *        <input
+ *          value={data.email}
+ *          onChange={({ target: { value }}) => setUrlState(s => ({
+ *            ...s,
+ *            data: { ...s.data, email: value },
+ *          }))}
+ *        />
+ *        <Button onClick={onClick}>{urlState.count} Increment</Button>
+ *        <hr />
+ *        State:
+ *        <pre>
+ *            {JSON.stringify(urlState, null, 2)}
+ *        </pre>
+ *     </div>
+ *   );
+ * }
+ * <ExampleUrlSyncControlled />
+ * ```
+ */
+export default function useUrlQueryState(
+    initialState: {} = {},
+    options: Options = {}
+): [
+    /**
+     * The current query state
+     *
+     * If `options.prefix` is specified this contains only query param keys that
+     * start with the prefix otherwise it contains _all_ query params.
+     */
+    Record<string, any>,
+    /**
+     * State transition function. Accepts either the state object to transition
```

Comment:
yes

Reply:
Can you make that more explicit please

Reply:
What do you suggest? the latest version since your original comment already includes the line 

> Passing the function returned by `useState` is compatible with this.

Reply:
Ah I missed that the change was at the end of the comment block, I thought you hadn't changed anything.

> Passing the function returned by

Is that a typo? Isn't it the transition function passed _into_ `useState()` rather than the state setting function returned from `useState()` that is compatible?


This makes sense to me:

```
    /**
     * State transition function. Accepts either the state object to transition
     * to OR a transition function.
     *
     * A transition function should accept the current state and return the new
     * state (similar to a `useState` transition function).
     *
     * Unlike `useState` the returned state always replaces the current state - it is not merged
     * except with the caveats noted below.
```

(Feel free to reword if I'm wrong or you think could be clearer)

Reply:
It's not correct - `useState` always replaces the state it never merges same as this

The only thing you (optionally) pass to `useState` is the initial state. It returns the state and the function to call to change state. So what I wrote was correct in that it's the return value of `useState` you pass to this